Revert commit 1ccd638b1aa85fb3c43b49d69d279cd509ebdc21
Author: Alexandre Julliard <julliard@winehq.org>
Date:   Tue Apr 21 11:27:53 2020 +0200

    winegcc: No longer use a constructor for module initialization.
    
    Signed-off-by: Alexandre Julliard <julliard@winehq.org>

--- UTC
diff --git tools/winegcc/winegcc.c tools/winegcc/winegcc.c
index 8a7db2ad34..9008c457bd 100644
--- tools/winegcc/winegcc.c
+++ tools/winegcc/winegcc.c
@@ -553,6 +553,8 @@ static strarray *get_link_args( struct options *opts, const char *output_name )
         strarray_add( flags, "-Wl,--no-wchar-size-warning" );
     if (!try_link( opts->prefix, link_args, "-Wl,-z,defs" ))
         strarray_add( flags, "-Wl,-z,defs" );
+    if (opts->shared && !try_link( opts->prefix, link_args, "-Wl,-init,__wine_spec_init" ))
+        strarray_add( flags, "-Wl,-init,__wine_spec_init" );
 
     strarray_addall( link_args, flags );
     return link_args;
@@ -1408,8 +1410,8 @@ static void build(struct options* opts)
         }
     }
 
-    if (!is_pe) fixup_constructors( opts, output_path );
-    else if (opts->wine_builtin) make_wine_builtin( opts, output_path );
+    if (!is_pe && !opts->shared) fixup_constructors( opts, output_path );
+    if (is_pe && opts->wine_builtin) make_wine_builtin( opts, output_path );
 
     /* create the loader script */
     if (generate_app_loader)
